rosbag2_transport/src/rosbag2_transport/player_service_client.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright 2023 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "rosbag2_transport/player_service_client.hpp"
+
+#include "rclcpp/logger.hpp"
+
+#include "rosbag2_cpp/service_utils.hpp"
+
+#include "rosidl_typesupport_introspection_cpp/identifier.hpp"
+#include "rosidl_typesupport_introspection_cpp/message_introspection.hpp"
+
+#include "logging.hpp"
+
+namespace rosbag2_transport
+{
+
+PlayerServiceClient::PlayerServiceClient(
+  std::shared_ptr<rclcpp::GenericClient> cli,
+  std::string service_name,
+  const std::string & service_event_type,
+  const rclcpp::Logger logger,
+  std::shared_ptr<PlayerServiceClientManager> player_service_client_manager)
+: client_(std::move(cli)),
+  service_name_(std::move(service_name)),
+  logger_(logger),
+  player_service_client_manager_(player_service_client_manager)
+{
+  ts_lib_ = rclcpp::get_typesupport_library(
+    service_event_type, "rosidl_typesupport_cpp");
+
+  ts_ = rclcpp::get_message_typesupport_handle(
+    service_event_type,
+    "rosidl_typesupport_cpp",
+    *ts_lib_);
+
+  auto message_ts_handle = get_message_typesupport_handle(
+    ts_,
+    rosidl_typesupport_introspection_cpp::typesupport_identifier);
+
+  message_members_ =
+    reinterpret_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+    message_ts_handle->data);
+}
+
+bool PlayerServiceClient::is_include_request_message(
+  const rclcpp::SerializedMessage & message)
+{
+  auto type = get_msg_event_type(message);
+
+  // Ignore response message
+  if (type == service_msgs::msg::ServiceEventInfo::RESPONSE_SENT ||
+    type == service_msgs::msg::ServiceEventInfo::RESPONSE_RECEIVED)
+  {
+    return false;
+  }
+
+  // Ignore metadata message
+  if ((type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT ||
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED) &&
+    client_side_type_ == introspection_type::METADATA)
+  {
+    return false;
+  }
+
+  if (client_side_type_ == introspection_type::UNKNOW &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      client_side_type_ = introspection_type::METADATA;
```

Comment:
isnt this bad assumption? this rosbag2 database could have multiple client's data in the same service name. i do not think we can cache the introspection type in generic client here unless we manage that for each client ID.

the assumption we can make here,

- we are not sure if server or clients enabled `contents` or `metadata`. some did, some did not.
- there are multiple clients to the service name, this is likely.
- we do not want to duplicated request from the same transaction with `REQUEST_SENT` and `REQUEST_RECEIVED`
- introspection flag can be changed dynamically, and user application would do so on server and clients.
- https://github.com/ros2/rmw/issues/357 does not allow us to tell the identical transaction.

how about the following for now with some note?

always try to send request based on service server data with `CONTENTS` is enabled. expectation is that we can guide the user to enable service introspection on service server side with this feature, and that is gonna be much easier to take advantage of this feature instead of enabling all clients.
when https://github.com/ros2/rmw/issues/357 solves, we can control more flexibly with client ID if the request is played or not as much as possible with server and client data, besides we can also avoid request duplication.

Reply:
> isnt this bad assumption? this rosbag2 database could have multiple client's data in the same service name.

Yes. I need to reconsider the logic of the code.

> always try to send request based on service server data with CONTENTS is enabled. expectation is that we can guide the user to enable service introspection on service server side with this feature, and that is gonna be much easier to take advantage of this feature instead of enabling all clients.

If introspection is set on the service side, send the received request sent from the service side. This can simplify the code logic. 

> when https://github.com/ros2/rmw/issues/357 solves, we can control more flexibly with client ID if the request is played or not as much as possible with server and client data, besides we can also avoid request duplication.

Yes. I will reconsider this part of the logic based on the client ID.

Reply:
After considering, we cannot use "always try to send request based on service server data with CONTENTS is enabled. ".  

e.g.   
In recorded data  
1. request_sent (for client 1)
2. request_received (for client 1)
3. ...

While dealing with 1, code doesn't know whether service enable introspection. But code must decide whether this request is sent this time. At this case, code have to decide to use request data from request_sent to send. Code have to do judgement for each client. So I change the logic of code https://github.com/ros2/rosbag2/pull/1481/commits/05c1e95d9ef6e153e165c59c467f0c4ec69e1d3a.